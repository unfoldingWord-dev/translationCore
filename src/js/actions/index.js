/**
 * @author Ian Hoegen
 * @description: This serves as the new store and events api.
 * @usage:
 * Actions.sendAction('YOUR_EVENT', data) // Emits an event of type YOUR_EVENT, adds data to store
 * Actions.getData('YOUR_EVENT') // Retrieves data from the event emitted
 ***********************************************************************************/

const EventEmitter = require('events').EventEmitter;
const Reducers = require('../reducers');

class Actions extends EventEmitter {
  constructor() {
    super();
    this.store = [];
  }
  sendAction(eventName, actionObj) {
    if(Reducers[eventName]) {
      this.store[eventName] = Reducers[eventName](this.store[eventName], actionObj);
    } else {
      this.store[eventName] = actionObj;
    }
    this.emitChange(eventName);
  }

  subscribe(eventName, callback) {
    this.on(eventName, callback);
  }

  unsubscribe(eventName, callback) {
    this.removeListener(eventName, callback);
  }

  emitChange(eventName, data) {
    this.emit(eventName, data);
  }

  getData (actionType) {
    return this.store[actionType];
  }
}

var actions = new Actions();
module.exports = actions;
