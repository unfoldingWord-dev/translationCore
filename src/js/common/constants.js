import path from "path-extra";
import ospath from "ospath";
import packagefile from '../../../package.json';

export const APP_VERSION = packagefile.version;
export const MIN_COMPATIBLE_VERSION = packagefile.minCompatibleVersion;

export const TRANSLATION_WORDS = 'translationWords';
export const TRANSLATION_NOTES = 'translationNotes';
export const TRANSLATION_ACADEMY = 'translationAcademy';
export const TRANSLATION_HELPS = 'translationHelps';
export const WORD_ALIGNMENT = 'wordAlignment';

export const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');
export const USER_RESOURCES_PATH = path.join(ospath.home(), "translationCore", "resources");
export const THELPS_EN_RESOURCES_PATH = path.join(USER_RESOURCES_PATH, 'en', TRANSLATION_HELPS);
export const IMPORTS_PATH = path.join(ospath.home(), 'translationCore', 'imports');
export const PROJECT_INDEX_FOLDER_PATH = path.join('.apps', 'translationCore', 'index');
export const TEMP_IMPORT_PATH = path.join(ospath.home(), 'translationCore', 'imports', 'temp');

export const PROJECT_DOT_APPS_PATH = path.join('.apps', 'translationCore');
export const PROJECT_CHECKDATA_DIRECTORY = path.join(PROJECT_DOT_APPS_PATH, 'checkData');
export const STATIC_RESOURCES_PATH = path.join(__dirname, "../../../tcResources");

export const TC_VERSION = "tc_version";
export const SOURCE_CONTENT_UPDATER_MANIFEST = "source-content-updater-manifest.json";

export const tc_EDIT_VERSION_KEY = "tc_edit_version"; // do not change this string as it will break compatibility with other versions
export const tc_MIN_COMPATIBLE_VERSION_KEY = "tc_min_compatible_version"; // do not change this string as it will break compatibility with other versions
export const tc_MIN_VERSION_ERROR = "TC_MIN_VERSION_ERROR";

export const ORIGINAL_LANGUAGE = "originalLanguage";
export const TARGET_LANGUAGE = "targetLanguage";
export const TARGET_BIBLE = "targetBible";

export const toolCardCategories = {
  'kt': 'Key Terms',
  'names': 'Names',
  'other_terms': 'Other Terms',

  'culture': 'Culture',
  'figures': 'Figures of Speech',
  'numbers': 'Numbers',
  'discourse': 'Discourse',
  'grammar': 'Grammar',
  'other': 'Other'
};
