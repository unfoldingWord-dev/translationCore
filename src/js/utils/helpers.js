import Path from 'path';
import * as fs from 'fs-extra';
import ManifestGenerator from '../components/core/create_project/ProjectManifest';
import BooksOfBible from '../components/core/BooksOfBible';

module.exports.isOldTestament = function (projectBook) {
    var passedBook = false;
    for (var book in BooksOfBible) {
        if (book == projectBook) passedBook = true;
        if (BooksOfBible[book] == "Malachi" && passedBook) {
            return true;
        }
    }
    return false;
}

module.exports.checkIfValidBetaProject = function (manifest) {
    if (manifest && manifest.project) return manifest.project.id == "eph" || manifest.project.id == "tit";
    else if (manifest && manifest.ts_project) return manifest.ts_project.id == "eph" || manifest.ts_project.id == "tit";
}

/**
 * 
 * @param {*} path 
 * @param {*} file 
 */
export function loadFile(path, file) {
    try {
        var manifest = fs.readJsonSync(Path.join(path, file));
        return manifest;
    }
    catch (e) {
        return null
    }
}

/**
 * @description - Generates and saves a translationCore manifest file
 * @param {string} projectSaveLocation - Filepath of where the translationCore manifest file will
 * be saved. Must be an ABSOLUTE PATH
 * @param {object} data - The translationCore manifest data to be saved
 * @param {object} tsManifest - The translationStudio manifest data loaded from a translation
 * studio project
 */
module.exports.saveManifest = function (projectSaveLocation, link, oldManifest, currentUser) {
    var data = {
        user: [currentUser],
        repo: link
    }
    var manifest;
    try {
        var manifestLocation = Path.join(projectSaveLocation, 'manifest.json');
        if (oldManifest.package_version == '3') {
            //some older versions of ts-manifest have to be tweaked to work
            manifest = this.fixManifestVerThree(oldManifest);
        } else {
            manifest = ManifestGenerator(data, oldManifest);
        }
        fs.outputJsonSync(manifestLocation, manifest);
    }
    catch (err) {
        console.error(err);
    }
    return manifest;
}

/**
 * @desription - Uses the tc-standard format for projects to make package_version 3 compatible
 * @param oldManifest - The name of an employee.
 */
module.exports.fixManifestVerThree = function (oldManifest) {
    var newManifest = {};
    try {
        for (var oldElements in oldManifest) {
            newManifest[oldElements] = oldManifest[oldElements];
        }
        newManifest.finished_chunks = oldManifest.finished_frames;
        newManifest.ts_project = {};
        newManifest.ts_project.id = oldManifest.project_id;
        newManifest.ts_project.name = this.convertToFullBookName(oldManifest.project_id);
        for (var el in oldManifest.source_translations) {
            newManifest.source_translations = oldManifest.source_translations[el];
            var parameters = el.split("-");
            newManifest.source_translations.language_id = parameters[1];
            newManifest.source_translations.resource_id = parameters[2];
            break;
        }
    } catch (e) {
        console.error(e);
    }
    return newManifest;
}

module.exports.convertToFullBookName = function (bookAbbr) {
    if (!bookAbbr) return;
    return BooksOfBible[bookAbbr.toString().toLowerCase()];
}