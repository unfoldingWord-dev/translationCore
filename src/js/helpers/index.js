import { generateTimestamp } from './TimestampGenerator.js';
import * as loadHelpers from './LoadHelpers.js';

export {
  generateTimestamp,
  loadHelpers,
};
