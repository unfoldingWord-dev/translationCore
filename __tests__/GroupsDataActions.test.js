/* eslint-disable object-curly-newline */
import fs from 'fs-extra';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import path from 'path-extra';
// actions
import { generateTimestamp } from '../src/js/helpers';
import * as GroupsDataActions from '../src/js/actions/GroupsDataActions';
import * as saveMethods from '../src/js/localStorage/saveMethods';
import { WORD_ALIGNMENT, TRANSLATION_WORDS } from '../src/js/common/constants';
// constants
const FIXTURES_CHECKDATA_PATH = path.join(__dirname, 'fixtures', 'checkData');
const CURRENT_PROJECT_PATH = path.join(__dirname, 'fixtures', 'project', 'en_tit');
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('redux-batched-actions', () => ({
  batchActions: (actionsBatch) => (dispatch) => {
    if (actionsBatch.length) {
      for (let action of actionsBatch) {
        dispatch(action);
      }
    }
  },
}));

describe('GroupsDataActions.verifyGroupDataMatchesWithFs', () => {
  let saveOtherContextSpy = null;

  beforeEach(() => {
    saveOtherContextSpy = jest.spyOn(saveMethods,
      'saveSelectionsForOtherContext');
    fs.__resetMockFS();
    fs.__loadDirIntoMockFs(FIXTURES_CHECKDATA_PATH, FIXTURES_CHECKDATA_PATH);
    fs.__loadDirIntoMockFs(CURRENT_PROJECT_PATH, CURRENT_PROJECT_PATH);
  });

  afterEach(() => {
    removeSpy(saveOtherContextSpy);
  });

  it('should succeed without external verse edits', async () => {
    // given
    const bookId = 'tit';
    const groupsDataReducer = fs.readJsonSync(path.join(FIXTURES_CHECKDATA_PATH, 'en_tit', 'groupsDataReducer.json'));
    const initStore = {
      groupsDataReducer,
      toolsReducer: { selectedTool: TRANSLATION_WORDS },
      projectDetailsReducer: {
        manifest: { project: { id: bookId } },
        projectSaveLocation: CURRENT_PROJECT_PATH,
      },
    };
    const store = mockStore(initStore);

    // when
    await store.dispatch(GroupsDataActions.verifyGroupDataMatchesWithFs(initStore.toolsReducer.selectedTool));

    // then
    const actions = store.getActions();
    expect(cleanOutDates(actions)).toMatchSnapshot();
  });

  it('should succeed with external verse edits', async () => {
    // given
    const bookId = 'tit';
    const groupsDataReducer = fs.readJsonSync(path.join(FIXTURES_CHECKDATA_PATH, 'en_tit', 'groupsDataReducer.json'));
    const initStore = {
      groupsDataReducer,
      toolsReducer: { selectedTool: TRANSLATION_WORDS },
      projectDetailsReducer: {
        manifest: { project: { id: bookId } },
        projectSaveLocation: CURRENT_PROJECT_PATH,
      },
    };
    const store = mockStore(initStore);
    // add verse edit
    const verseEdit = {
      'verseBefore':'To Titus, a true son in our common faith. Grace and peace from God the Father and Christ Jesus our savior.\n\\p','verseAfter':'To Titus, a true son in our common faith. Grace and peace from God the Father and Christ Jesus our savior.\n\\p Edit 1:4','tags':['other'],'userName':'photonomad1','activeBook':'tit','activeChapter':1,'activeVerse':1,'modifiedTimestamp':'2019-05-16T12:11:45.970Z','gatewayLanguageCode':'en','gatewayLanguageQuote':'','contextId':{
        'reference':{
          'bookId':'tit','chapter':1,'verse':4,
        },'tool':WORD_ALIGNMENT,'groupId':'chapter_1',
      },
    };
    const verseEditPath = path.join(CURRENT_PROJECT_PATH, '.apps/translationCore/checkData', 'verseEdits', bookId, '1', '4');
    fs.ensureDirSync(verseEditPath);
    const fileName = generateTimestamp() + '.json';
    fs.outputJsonSync(path.join(verseEditPath, fileName), verseEdit);

    // when
    await store.dispatch(GroupsDataActions.verifyGroupDataMatchesWithFs(initStore.toolsReducer.selectedTool));

    // then
    const actions = store.getActions();
    expect(cleanOutDates(actions)).toMatchSnapshot();
  });

  it('should succeed with multiple external verse edits', async () => {
    // given
    const bookId = 'tit';
    const groupsDataReducer = fs.readJsonSync(path.join(FIXTURES_CHECKDATA_PATH, 'en_tit', 'groupsDataReducer.json'));
    const initStore = {
      groupsDataReducer,
      toolsReducer: { selectedTool: TRANSLATION_WORDS },
      projectDetailsReducer: {
        manifest: { project: { id: bookId } },
        projectSaveLocation: CURRENT_PROJECT_PATH,
      },
    };
    const store = mockStore(initStore);
    // add verse edit
    let verseEdit = {
      'verseBefore':'To Titus, a true son in our common faith. Grace and peace from God the Father and Christ Jesus our savior.\n\\p','verseAfter':'To Titus, a true son in our common faith. Grace and peace from God the Father and Christ Jesus our savior.\n\\p Edit 1:4','tags':['other'],'userName':'photonomad1','activeBook':'tit','activeChapter':1,'activeVerse':1,'modifiedTimestamp':'2019-05-16T12:11:45.970Z','gatewayLanguageCode':'en','gatewayLanguageQuote':'','contextId':{
        'reference':{
          'bookId':'tit','chapter':1,'verse':4,
        },'tool':WORD_ALIGNMENT,'groupId':'chapter_1',
      },
    };
    let verseEditPath = path.join(CURRENT_PROJECT_PATH, '.apps/translationCore/checkData', 'verseEdits', bookId, '1', '4');
    fs.ensureDirSync(verseEditPath);
    let fileName = generateTimestamp() + '.json';
    fs.outputJsonSync(path.join(verseEditPath, fileName), verseEdit);

    // add 2nd verse edit
    verseEdit = {
      'verseBefore':'To Titus, a true son in our common faith. Grace and peace from God the Father and Christ Jesus our savior.\n\\p','verseAfter':'To Titus, a true son in our common faith. Grace and peace from God the Father and Christ Jesus our savior.\n\\p Edit 1:7','tags':['other'],'userName':'photonomad1','activeBook':'tit','activeChapter':1,'activeVerse':7,'modifiedTimestamp':'2019-05-16T12:11:45.970Z','gatewayLanguageCode':'en','gatewayLanguageQuote':'','contextId':{
        'reference':{
          'bookId':'tit','chapter':1,'verse':7,
        },'tool':WORD_ALIGNMENT,'groupId':'chapter_1',
      },
    };
    verseEditPath = path.join(CURRENT_PROJECT_PATH, '.apps/translationCore/checkData', 'verseEdits', bookId, '1', '7');
    fs.ensureDirSync(verseEditPath);
    fileName = generateTimestamp() + '.json';
    fs.outputJsonSync(path.join(verseEditPath, fileName), verseEdit);

    // when
    await store.dispatch(GroupsDataActions.verifyGroupDataMatchesWithFs(initStore.toolsReducer.selectedTool));

    // then
    const actions = store.getActions();
    expect(cleanOutDates(actions)).toMatchSnapshot();
  });
});

describe('GroupsDataActions.validateBookSelections', () => {
  const bookId = 'tit';
  const selectionsReducer = {
    'gatewayLanguageCode': 'en',
    'gatewayLanguageQuote': 'authority, authorities',
    'selections': [
      {
        'text': 'apostle',
        'occurrence': 1,
        'occurrences': 1,
      },
    ],
    'username': 'dummy-test',
    'modifiedTimestamp': generateTimestamp(),
  };
  let saveOtherContextSpy = null;

  beforeEach(() => {
    saveOtherContextSpy = jest.spyOn(saveMethods,
      'saveSelectionsForOtherContext');
    fs.__resetMockFS();
    fs.__loadDirIntoMockFs(FIXTURES_CHECKDATA_PATH, FIXTURES_CHECKDATA_PATH);
    fs.__loadDirIntoMockFs(CURRENT_PROJECT_PATH, CURRENT_PROJECT_PATH);
  });

  afterEach(() => {
    removeSpy(saveOtherContextSpy);
  });

  it('No selection changes', () => {
    // given
    const store = initiMockStore(bookId, selectionsReducer);
    const expectedSelectionChanges = 0;

    // when
    store.dispatch(GroupsDataActions.validateBookSelections());

    // then
    const actions = store.getActions();
    expect(cleanOutDates(actions)).toMatchSnapshot();
    expect(saveOtherContextSpy).toHaveBeenCalledTimes(expectedSelectionChanges);
  });

  it('apostle selection edited', () => {
    // given
    const targetVerse = 'Paul, a servant of God and an apostl2 of Jesus Christ, for the faith of God\'s chosen people and the knowledge of the truth that agrees with godliness, ';
    const store = initiMockStore(bookId, selectionsReducer, '1', '1',
      targetVerse);
    const expectedSelectionChanges = 1;

    // when
    store.dispatch(GroupsDataActions.validateBookSelections());

    // then
    const actions = store.getActions();
    expect(cleanOutDates(actions)).toMatchSnapshot();
    expect(saveOtherContextSpy).toHaveBeenCalledTimes(expectedSelectionChanges);
  });

  it('all selections edited', () => {
    // given
    const targetVerse = '';
    const store = initiMockStore(bookId, selectionsReducer, '1', '1',
      targetVerse);
    const expectedSelectionChanges = 2;

    // when
    store.dispatch(GroupsDataActions.validateBookSelections());

    // then
    const actions = store.getActions();
    expect(cleanOutDates(actions)).toMatchSnapshot();
    expect(saveOtherContextSpy).toHaveBeenCalledTimes(expectedSelectionChanges);
  });
});

describe('GroupsDataActions.isAttributeChanged', () => {
  describe('reminders', () => {
    it('new value true and old value true is unchanged', () => {
      const newValue = true;
      const oldValue = true;
      const changed = false;
      validateRemindersChanged(newValue, oldValue, changed);
    });
    it('new value false and old value false is unchanged', () => {
      const newValue = false;
      const oldValue = false;
      const changed = false;
      validateRemindersChanged(newValue, oldValue, changed);
    });
    it('new value true and old value false is changed', () => {
      const newValue = true;
      const oldValue = false;
      const changed = true;
      validateRemindersChanged(newValue, oldValue, changed);
    });
    it('new value false and old value undefined is unchanged', () => {
      const newValue = false;
      const oldValue = undefined;
      const changed = false;
      validateRemindersChanged(newValue, oldValue, changed);
    });
    it('new value false and old value null is unchanged', () => {
      const newValue = false;
      const oldValue = null;
      const changed = false;
      validateRemindersChanged(newValue, oldValue, changed);
    });
  });
  describe('invalidated', () => {
    it('new value true and old value true is unchanged', () => {
      const newValue = true;
      const oldValue = true;
      const changed = false;
      validateInvalidationChanged(newValue, oldValue, changed);
    });
    it('new value false and old value false is unchanged', () => {
      const newValue = false;
      const oldValue = false;
      const changed = false;
      validateInvalidationChanged(newValue, oldValue, changed);
    });
    it('new value true and old value false is changed', () => {
      const newValue = true;
      const oldValue = false;
      const changed = true;
      validateInvalidationChanged(newValue, oldValue, changed);
    });
    it('new value false and old value undefined is unchanged', () => {
      const newValue = false;
      const oldValue = undefined;
      const changed = false;
      validateInvalidationChanged(newValue, oldValue, changed);
    });
    it('new value false and old value null is unchanged', () => {
      const newValue = false;
      const oldValue = null;
      const changed = false;
      validateInvalidationChanged(newValue, oldValue, changed);
    });
  });
  describe('comments', () => {
    it('new value string same as old value is unchanged', () => {
      const newValue = 'stuff';
      const oldValue = 'stuff';
      const changed = false;
      validateCommentsChanged(newValue, oldValue, changed);
    });
    it('new value string different than old value string is changed', () => {
      const newValue = 'stuff';
      const oldValue = 'stuff2';
      const changed = true;
      validateCommentsChanged(newValue, oldValue, changed);
    });
    it('new value string and old value true is changed', () => {
      const newValue = 'stuff';
      const oldValue = true;
      const changed = true;
      validateCommentsChanged(newValue, oldValue, changed);
    });
    it('new value string and old value false is changed', () => {
      const newValue = 'stuff';
      const oldValue = false;
      const changed = true;
      validateCommentsChanged(newValue, oldValue, changed);
    });
    it('new value empty string and old value false is unchanged', () => {
      const newValue = '';
      const oldValue = false;
      const changed = false;
      validateCommentsChanged(newValue, oldValue, changed);
    });
    it('new value empty string and old value string is changed', () => {
      const newValue = '';
      const oldValue = 'stuffff';
      const changed = true;
      validateCommentsChanged(newValue, oldValue, changed);
    });
    it('new value empty string and old value empty string is unchanged', () => {
      const newValue = '';
      const oldValue = '';
      const changed = false;
      validateCommentsChanged(newValue, oldValue, changed);
    });
    it('new value null and old value false is unchanged', () => {
      const newValue = null;
      const oldValue = false;
      const changed = false;
      validateCommentsChanged(newValue, oldValue, changed);
    });
    it('new value false and old value null is unchanged', () => {
      const newValue = false;
      const oldValue = null;
      const changed = false;
      validateCommentsChanged(newValue, oldValue, changed);
    });
  });
  describe('selections', () => {
    it('new value selection same as old value is unchanged', () => {
      const newSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 } ];
      const oldSelectValue = [ { occurrence: 1, occurrences: 2, text: 'Jacob' } ];
      const changed = false;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value selection different than old value selection is changed', () => {
      const newSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 } ];
      const oldSelectValue = [ { text: 'Jacob', occurrence: 2, occurrences: 2 } ];
      const changed = true;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value selection different count than old value selection is changed', () => {
      const newSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 } ];
      const oldSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 }, { text: 'son', occurrence: 1, occurrences: 4 } ];
      const changed = true;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value selection different order than old value selection is unchanged', () => {
      const newSelectValue = [ { text: 'son', occurrence: 1, occurrences: 4 }, { text: 'Jacob', occurrence: 1, occurrences: 2 } ];
      const oldSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 }, { text: 'son', occurrence: 1, occurrences: 4 } ];
      const changed = true;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value selection and old value true is changed', () => {
      const newSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 } ];
      const oldSelectValue = true;
      const changed = true;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value selection and old value false is changed', () => {
      const newSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 } ];
      const oldSelectValue = false;
      const changed = true;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value empty selection and old value false is unchanged', () => {
      const newSelectValue = [];
      const oldSelectValue = false;
      const changed = false;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value false and old value empty selection is unchanged', () => {
      const newSelectValue = false;
      const oldSelectValue = [];
      const changed = false;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value empty selection and old value selection is changed', () => {
      const newSelectValue = [];
      const oldSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 } ];
      const changed = true;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value null and old value selection is changed', () => {
      const newSelectValue = null;
      const oldSelectValue = [ { text: 'Jacob', occurrence: 1, occurrences: 2 } ];
      const changed = true;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value empty selection and old value empty selection is unchanged', () => {
      const newSelectValue = [];
      const oldSelectValue = [];
      const changed = false;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value null and old value false is unchanged', () => {
      const newSelectValue = null;
      const oldSelectValue = false;
      const changed = false;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value undefined and old value false is unchanged', () => {
      const newSelectValue = undefined;
      const oldSelectValue = false;
      const changed = false;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
    it('new value false and old value null is unchanged', () => {
      const newSelectValue = false;
      const oldSelectValue = null;
      const changed = false;
      validateSelectionsChanged(newSelectValue, oldSelectValue, changed);
    });
  });
  describe('nothingToSelect', () => {
    it('new nothingToSelect value true and old value true is unchanged', () => {
      const newNoSelectValue = true;
      const oldNoSelectValue = true;
      const changed = false;
      validateNothingToSelectChanged(newNoSelectValue, oldNoSelectValue, changed);
    });
    it('new nothingToSelect value false and old value false is unchanged', () => {
      const newNoSelectValue = false;
      const oldNoSelectValue = false;
      const changed = false;
      validateNothingToSelectChanged(newNoSelectValue, oldNoSelectValue, changed);
    });
    it('new nothingToSelect value true and old value false is changed', () => {
      const newNoSelectValue = true;
      const oldNoSelectValue = false;
      const changed = true;
      validateNothingToSelectChanged(newNoSelectValue, oldNoSelectValue, changed);
    });
    it('new nothingToSelect value false and old value true is changed', () => {
      const newNoSelectValue = false;
      const oldNoSelectValue = true;
      const changed = true;
      validateNothingToSelectChanged(newNoSelectValue, oldNoSelectValue, changed);
    });
    it('new value false and old value null is unchanged', () => {
      const newNoSelectValue = false;
      const oldNoSelectValue = null;
      const changed = false;
      validateNothingToSelectChanged(newNoSelectValue, oldNoSelectValue, changed);
    });
    it('new value null and old value false is unchanged', () => {
      const newNoSelectValue = null;
      const oldNoSelectValue = false;
      const changed = false;
      validateNothingToSelectChanged(newNoSelectValue, oldNoSelectValue, changed);
    });
  });
  describe('verseEdits', () => {
    it('old value true is unchanged', () => {
      const oldValue = true;
      const changed = false;
      validateVerseEdits(oldValue, changed);
    });
    it('old value false is changed', () => {
      const oldValue = false;
      const changed = true;
      validateVerseEdits(oldValue, changed);
    });
    it('old value null is changed', () => {
      const oldValue = null;
      const changed = true;
      validateVerseEdits(oldValue, changed);
    });
  });
  describe('unsupported', () => {
    it('unsupported checks are always false', () => {
      const newValue = true;
      const oldValue = null;
      const expectChange = false;
      const checkAttr = 'unsupported';
      const object = { [checkAttr]: newValue };
      const oldGroupObject = { [checkAttr]: oldValue };

      // when
      const result = GroupsDataActions.isAttributeChanged(object, checkAttr, oldGroupObject);

      // then
      expect(result === expectChange).toBeTruthy();
    });
  });
});

//
// helpers
//

function validateVerseEdits(oldValue, expectChange) {
  const checkAttr = 'verseEdits';
  const object = {};
  const oldGroupObject = { [checkAttr]: oldValue };
  const result = GroupsDataActions.isAttributeChanged(object, checkAttr, oldGroupObject);
  expect(result === expectChange).toBeTruthy();
}

function validateCommentsChanged(newValue, oldValue, expectChange) {
  const checkAttr = 'comments';
  const attr = 'text';
  const object = { [attr]: newValue };
  const oldGroupObject = { [checkAttr]: oldValue };
  const result = GroupsDataActions.isAttributeChanged(object, checkAttr, oldGroupObject);
  expect(result === expectChange).toBeTruthy();
}

function validateInvalidationChanged(newValue, oldValue, expectChange) {
  const checkAttr = 'invalidated';
  const object = { [checkAttr]: newValue };
  const oldGroupObject = { [checkAttr]: oldValue };
  const result = GroupsDataActions.isAttributeChanged(object, checkAttr, oldGroupObject);
  expect(result === expectChange).toBeTruthy();
}

function validateSelectionsChanged(newSelectValue, oldSelectValue, expectChange, newNoSelectValue = undefined, oldNoSelectValue= false) {
  const checkAttr = 'selections';
  const object = { [checkAttr]: newSelectValue, ['nothingToSelect']: newNoSelectValue };
  const oldGroupObject = { [checkAttr]: oldSelectValue, ['nothingToSelect']: oldNoSelectValue };
  const result = GroupsDataActions.isAttributeChanged(object, checkAttr, oldGroupObject);
  expect(result === expectChange).toBeTruthy();
}

function validateNothingToSelectChanged(newNoSelectValue, oldNoSelectValue, expectChange) {
  validateSelectionsChanged(null, false, expectChange, newNoSelectValue, oldNoSelectValue);
}

function validateRemindersChanged(newValue, oldValue, expectChange) {
  const checkAttr = 'reminders';
  const attr = 'enabled';
  const object = { [attr]: newValue };
  const oldGroupObject = { [checkAttr]: oldValue };
  const result = GroupsDataActions.isAttributeChanged(object, checkAttr, oldGroupObject);
  expect(result === expectChange).toBeTruthy();
}

function cleanOutDates(actions) {
  const cleanedActions = JSON.parse(JSON.stringify(actions));

  for (let action of cleanedActions) {
    if (action.modifiedTimestamp) {
      delete action.modifiedTimestamp;
    }
  }
  return cleanedActions;
}

function getInitialStateData(bookId, checkPath, projectPath) {
  const contextId = {
    reference: {
      bookId: bookId,
      chapter: 1,
      verse: 1,
    },
    groupId: '',
  };
  const groupsDataReducer = fs.readJSONSync(
    path.join(checkPath, 'groupsDataReducer.json'));
  const groupsIndexReducer = fs.readJSONSync(
    path.join(checkPath, 'groupsIndexReducer.json'));
  const targetBible = {
    1: fs.readJSONSync(path.join(projectPath, '1.json')),
    2: fs.readJSONSync(path.join(projectPath, '2.json')),
    3: fs.readJSONSync(path.join(projectPath, '3.json')),
  };

  const initialState = {
    actions: {},
    groupsDataReducer,
    groupsIndexReducer,
    projectInformationCheckReducer: { bookId },
    resourcesReducer: { bibles: { targetLanguage: { targetBible } } },
    toolsReducer: { selectedTool: TRANSLATION_WORDS },
    loginReducer: {
      loggedInUser: false,
      userdata: { username: 'dummy-test' },
    },
    projectDetailsReducer: {
      manifest: {
        project: { id: bookId },
        toolsSelectedGLs: { translationWords: 'en' },
      },
      projectSaveLocation: path.resolve(checkPath),
      currentToolName: TRANSLATION_WORDS,
    },
    contextIdReducer: { contextId },
  };
  return initialState;
}

function removeSpy(spy) {
  if (spy) {
    spy.mockReset();
    spy.mockRestore();
  }
}

function initiMockStore(bookId, selectionsReducer, chapter = null, verse = null, targetVerse = null) {
  const checkPath = path.join(FIXTURES_CHECKDATA_PATH, 'en_tit');
  const projectPath = path.join(CURRENT_PROJECT_PATH, 'tit');
  const initialState = getInitialStateData(bookId, checkPath, projectPath);
  initialState.selectionsReducer = selectionsReducer;
  initialState.projectDetailsReducer = {
    ...initialState.projectDetailsReducer,
    projectSaveLocation: CURRENT_PROJECT_PATH,
    manifest: {
      ...initialState.projectDetailsReducer.manifest,
      project: { id: bookId },
    },
  };

  if (typeof targetVerse === 'string') {
    const bibleChapter = initialState.resourcesReducer.bibles.targetLanguage.targetBible[chapter];
    bibleChapter[verse] = targetVerse;
  }
  return mockStore(initialState);
}

