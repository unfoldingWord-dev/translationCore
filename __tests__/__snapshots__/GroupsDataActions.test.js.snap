// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupsDataActions.validateBookSelections No selection changes 1`] = `Array []`;

exports[`GroupsDataActions.validateBookSelections all selections edited 1`] = `
Array [
  Object {
    "contextId": Object {
      "groupId": "apostle",
      "occurrence": 1,
      "quote": "ἀπόστολος",
      "reference": Object {
        "bookId": "tit",
        "chapter": 1,
        "verse": 1,
      },
      "strong": Array [
        "G06520",
      ],
      "tool": "translationWords",
    },
    "selections": Array [],
    "type": "TOGGLE_SELECTIONS_IN_GROUPDATA",
  },
  Object {
    "boolean": true,
    "contextId": Object {
      "groupId": "apostle",
      "occurrence": 1,
      "quote": "ἀπόστολος",
      "reference": Object {
        "bookId": "tit",
        "chapter": 1,
        "verse": 1,
      },
      "strong": Array [
        "G06520",
      ],
      "tool": "translationWords",
    },
    "type": "SET_INVALIDATION_IN_GROUPDATA",
  },
  Object {
    "contextId": Object {
      "groupId": "god",
      "occurrence": 1,
      "quote": "Θεοῦ",
      "reference": Object {
        "bookId": "tit",
        "chapter": 1,
        "verse": 1,
      },
      "strong": Array [
        "G23160",
      ],
      "tool": "translationWords",
    },
    "selections": Array [],
    "type": "TOGGLE_SELECTIONS_IN_GROUPDATA",
  },
  Object {
    "boolean": true,
    "contextId": Object {
      "groupId": "god",
      "occurrence": 1,
      "quote": "Θεοῦ",
      "reference": Object {
        "bookId": "tit",
        "chapter": 1,
        "verse": 1,
      },
      "strong": Array [
        "G23160",
      ],
      "tool": "translationWords",
    },
    "type": "SET_INVALIDATION_IN_GROUPDATA",
  },
  Object {
    "alertMessage": "tools.selections_invalidated",
    "type": "OPEN_ALERT_DIALOG",
  },
]
`;

exports[`GroupsDataActions.validateBookSelections apostle selection edited 1`] = `
Array [
  Object {
    "contextId": Object {
      "groupId": "apostle",
      "occurrence": 1,
      "quote": "ἀπόστολος",
      "reference": Object {
        "bookId": "tit",
        "chapter": 1,
        "verse": 1,
      },
      "strong": Array [
        "G06520",
      ],
      "tool": "translationWords",
    },
    "selections": Array [],
    "type": "TOGGLE_SELECTIONS_IN_GROUPDATA",
  },
  Object {
    "boolean": true,
    "contextId": Object {
      "groupId": "apostle",
      "occurrence": 1,
      "quote": "ἀπόστολος",
      "reference": Object {
        "bookId": "tit",
        "chapter": 1,
        "verse": 1,
      },
      "strong": Array [
        "G06520",
      ],
      "tool": "translationWords",
    },
    "type": "SET_INVALIDATION_IN_GROUPDATA",
  },
  Object {
    "alertMessage": "tools.selections_invalidated",
    "type": "OPEN_ALERT_DIALOG",
  },
]
`;
